File: crates/dust_ir/src/serde.rs
use crate::{AnnotationValueIr, ConstructorParamIr};

/// Normalized rename strategies derived from `SerDeRename`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum SerdeRenameRuleIr {
    /// Lowercase words without separators.
    LowerCase,
    /// Uppercase words without separators.
    UpperCase,
    /// `PascalCase`.
    PascalCase,
    /// `camelCase`.
    CamelCase,
    /// `snake_case`.
    SnakeCase,
    /// `SCREAMING_SNAKE_CASE`.
    ScreamingSnakeCase,
    /// `kebab-case`.
    KebabCase,
    /// `SCREAMING-KEBAB-CASE`.
    ScreamingKebabCase,
}

/// Normalized serde-related configuration attached to a class.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct SerdeClassConfigIr {
    /// Optional explicit rename for the target.
    pub rename: Option<String>,
    /// Optional global field renaming rule.
    pub rename_all: Option<SerdeRenameRuleIr>,
    /// Internal or adjacent tag field for sealed class variants.
    pub tag: Option<String>,
    /// Adjacent content field for sealed class variants.
    pub content: Option<String>,
    /// Whether sealed class decoding should try variants without a tag.
    pub untagged: bool,
    /// Whether unknown keys should be rejected when deserializing.
    pub disallow_unrecognized_keys: bool,
    /// Sealed class variant metadata, in factory declaration order.
    pub variants: Vec<SerdeVariantConfigIr>,
}

impl SerdeClassConfigIr {
    /// Returns `true` when the config carries no effective settings.
    pub fn is_empty(&self) -> bool {
        self.rename.is_none()
            && self.rename_all.is_none()
            && self.tag.is_none()
            && self.content.is_none()
            && !self.untagged
            && !self.disallow_unrecognized_keys
            && self.variants.is_empty()
    }

    /// Returns whether the class config requests sealed class serde metadata.
    pub fn uses_sealed_representation(&self) -> bool {
        self.tag.is_some() || self.content.is_some() || self.untagged
    }
}

/// Normalized metadata for one sealed class SerDe variant.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SerdeVariantConfigIr {
    /// Source factory constructor name, for example `paid`.
    pub constructor_name: String,
    /// Redirected target class name, for example `PaymentPaid`.
    pub target_class_name: String,
    /// Resolved tag value after variant rename and class `renameAll`.
    pub tag: String,
    /// Source factory constructor parameters used to synthesize variant classes.
    pub params: Vec<ConstructorParamIr>,
}

/// Normalized serde-related configuration attached to one field.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct SerdeFieldConfigIr {
    /// Optional explicit JSON key rename.
    pub rename: Option<String>,
    /// Alternative accepted JSON keys during deserialization.
    pub aliases: Vec<String>,
    /// Raw source expression for one custom serde codec, if one was provided.
    pub codec_source: Option<String>,
    /// Raw source expression for a default value, if one was provided.
    pub default_value_source: Option<String>,
    /// Parser-owned typed default value, when available from annotation IR.
    pub default_value: Option<AnnotationValueIr>,
    /// Whether serialization should skip this field.
    pub skip_serializing: bool,
    /// Whether deserialization should skip this field.
    pub skip_deserializing: bool,
}

impl SerdeFieldConfigIr {
    /// Returns `true` when the config carries no effective settings.
    pub fn is_empty(&self) -> bool {
        self.rename.is_none()
            && self.aliases.is_empty()
            && self.codec_source.is_none()
            && self.default_value_source.is_none()
            && self.default_value.is_none()
            && !self.skip_serializing
            && !self.skip_deserializing
    }
}

File: packages/dust_dart/lib/src/serde/serde.dart
// coverage:ignore-file

import '../derive/base.dart';

/// Generates `serialize()` and `toJson()` support for the annotated class or enum.
final class Serialize extends DeriveTrait {
  /// Creates the `Serialize` derive marker.
  const Serialize();
}

/// Generates `_$TypeDeserialize(...)` and `_$TypeFromJson(...)` support.
final class Deserialize extends DeriveTrait {
  /// Creates the `Deserialize` derive marker.
  const Deserialize();
}

/// Instance contract generated for classes deriving [Serialize].
///
/// Dust uses [serialize] as the source verb. [toJson] is kept as the Dart
/// ecosystem mirror.
abstract interface class Serializable {
  /// Converts this value into a JSON object.
  Map<String, Object?> serialize();

  /// Dart ecosystem mirror for [serialize].
  Map<String, Object?> toJson();
}

/// Directional conversion contract for generated and custom serializers.
abstract interface class Serializer<DartT, JsonT> {
  /// Converts [value] into its JSON representation.
  JsonT serialize(DartT value);
}

/// Directional conversion contract for generated and custom deserializers.
abstract interface class Deserializer<DartT, JsonT> {
  /// Converts [json] into a Dart value.
  DartT deserialize(JsonT json);
}

/// Field-level conversion contract for custom Dust serde handling.
///
/// Dust owns nullability around the field. Codecs only serialize and
/// deserialize the non-null value itself.
abstract interface class SerDeCodec<DartT, JsonT>
    implements Serializer<DartT, JsonT>, Deserializer<DartT, JsonT> {
  /// Creates one codec contract object.
  const SerDeCodec(); // coverage:ignore-line
}

/// Dart ecosystem mirrors for serializers.
extension SerializerJsonMirror<DartT, JsonT> on Serializer<DartT, JsonT> {
  /// Dart ecosystem mirror for [Serializer.serialize].
  JsonT toJson(DartT value) => serialize(value);
}

/// Dart ecosystem mirrors for deserializers.
extension DeserializerJsonMirror<DartT, JsonT> on Deserializer<DartT, JsonT> {
  /// Dart ecosystem mirror for [Deserializer.deserialize].
  DartT fromJson(JsonT json) => deserialize(json);
}

/// The rename strategy used when Dust derives JSON keys automatically.
///
/// These values intentionally mirror the common Rust `serde(rename_all = "...")`
/// naming families, but use Dart enum-friendly casing.
enum SerDeRename {
  /// Convert names to `lowercase`.
  lowerCase,

  /// Convert names to `UPPERCASE`.
  upperCase,

  /// Convert names to `PascalCase`.
  pascalCase,

  /// Convert names to `camelCase`.
  camelCase,

  /// Convert names to `snake_case`.
  snakeCase,

  /// Convert names to `SCREAMING_SNAKE_CASE`.
  screamingSnakeCase,

  /// Convert names to `kebab-case`.
  kebabCase,

  /// Convert names to `SCREAMING-KEBAB-CASE`.
  screamingKebabCase,
}

/// Configures how Dust generates or interprets serde metadata.
///
/// This is intentionally modeled after Rust's `#[serde(...)]` attribute:
/// the same annotation can be placed on a declaration or on a field.
///
/// Common declaration-level options:
/// - [renameAll]
/// - [tag]
/// - [content]
/// - [untagged]
/// - [disallowUnrecognizedKeys]
///
/// Common field-level options:
/// - [rename]
/// - [defaultValue]
/// - [skip]
/// - [skipSerializing]
/// - [skipDeserializing]
/// - [aliases]
/// - [using]
final class SerDe extends DeriveConfig {
  /// Explicit serde rename for the annotated declaration or field.
  final String? rename;

  /// Automatic rename rule applied to child fields or enum variants.
  final SerDeRename? renameAll;

  /// JSON discriminator field for sealed class variants.
  ///
  /// On a sealed class, redirecting factory constructors define the variants;
  /// Dust generates omitted concrete target classes into the `.g.dart` part.
  final String? tag;

  /// JSON payload field for adjacent-tagged sealed class variants.
  final String? content;

  /// Whether sealed class deserialization should try variants without a tag.
  final bool untagged;

  /// Fallback value used when deserialization omits the annotated field.
  ///
  /// Fields skipped for deserialization with [skip] or [skipDeserializing]
  /// must provide this value so generated constructors stay total.
  final Object? defaultValue;

  /// Whether the field should be skipped for both serialization and
  /// deserialization.
  ///
  /// When [disallowUnrecognizedKeys] is enabled, skipped fields are not
  /// accepted as input keys.
  final bool skip;

  /// Whether the field should be skipped only for serialization.
  ///
  /// The field is still read during deserialization.
  final bool skipSerializing;

  /// Whether the field should be skipped only for deserialization.
  ///
  /// When [disallowUnrecognizedKeys] is enabled, this field is not accepted as
  /// an input key.
  final bool skipDeserializing;

  /// Alternate accepted input names for one field during deserialization.
  final List<String> aliases;

  /// Custom field codec used instead of Dust's built-in serde mapping.
  ///
  /// This should usually be a const object that implements [SerDeCodec]. Dust
  /// handles nullable fields outside the codec; codecs convert non-null values.
  final Object? using;

  /// Whether generated deserialization should reject unknown JSON keys on the
  /// annotated declaration.
  ///
  /// Primary field keys and [aliases] are recognized. Fields skipped for
  /// deserialization with [skip] or [skipDeserializing] are not accepted as
  /// input keys.
  final bool disallowUnrecognizedKeys;

  /// Creates one serde configuration annotation.
  const SerDe({
    this.rename,
    this.renameAll,
    this.tag,
    this.content,
    this.untagged = false,
    this.defaultValue,
    this.skip = false,
    this.skipSerializing = false,
    this.skipDeserializing = false,
    this.aliases = const [],
    this.using,
    this.disallowUnrecognizedKeys = false,
  });
}
